lib/pbench/test/unit/server/test_datasets_metadata.py

```diff
@@ -214,94 +205,38 @@ def test_get_bad_query_2(self, query_get_as):
-        assert response.status_code == HTTPStatus.BAD_REQUEST
-        assert response.json.get("message") == "Invalid request payload"
+        response = client.put(f"{server_config.rest_uri}/datasets/metadata/")
+        assert response.status_code == HTTPStatus.NOT_FOUND
 
-    def test_put_missing_keys(self, client, server_config, pbench_token):
+    def test_put_missing_key(self, client, server_config, pbench_token):
         """
         Test behavior when JSON payload does not contain all required keys.
 
         Note that Pbench will silently ignore any additional keys that are
         specified but not required.
-
-        TODO: This is mostly copied from commons.py; ideally these would be
-        refactored into a common "superclass" analagous to ApiBase as Commons
-        is to ElasticBase.
         """
-        classobject = DatasetsMetadata(client.config, client.logger)
-
-        def missing_key_helper(keys):
-            response = client.post(
-                f"{server_config.rest_uri}/datasets/metadata",
-                json=keys,
-            )
-            assert response.status_code == HTTPStatus.BAD_REQUEST
-            missing = sorted(set(required_keys) - set(keys))
-            assert (
-                response.json.get("message")
-                == f"Missing required parameters: {','.join(missing)}"
-            )
-
-        parameter_items = classobject.schema.parameters.items()
-
-        required_keys = [
-            key for key, parameter in parameter_items if parameter.required
-        ]
-
-        all_combinations = []
-        for r in range(1, len(parameter_items) + 1):
-            for item in itertools.combinations(parameter_items, r):
-                tmp_req_keys = [key for key, parameter in item if parameter.required]
-                if tmp_req_keys != required_keys:
-                    all_combinations.append(item)
-
-        for items in all_combinations:
-            keys = {}
-            for key, parameter in items:
-                if parameter.type is ParamType.ACCESS:
-                    keys[key] = "public"
-                elif parameter.type is ParamType.DATE:
-                    keys[key] = "2020"
-                elif parameter.type is ParamType.LIST:
-                    keys[key] = []
-                elif parameter.type is ParamType.KEYWORD:
-                    keys[key] = parameter.keywords[0]
-                else:
-                    keys[key] = "foobar"
-
-            missing_key_helper(keys)
-
-        # Test in case all of the required keys are missing and some
-        # random non-existent key is present in the payload
-        if required_keys:
-            missing_key_helper({"notakey": None})
+        response = client.put(
+            f"{server_config.rest_uri}/datasets/metadata/drb", json={}
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json.get("message") == "Missing required parameters: metadata"
```

Comment:
Does that message ("Missing required parameters: metadata") make sense to you in this context?  The request provided the metadata _resource_ ("drb"), but it didn't provide any metadata _key/value pairs_...so the token "metadata" seems, ambiguous at best (but, I don't know what to suggest, so I hope it's good enough as is...).

And, it's it really a bad request to ask to change nothing?

Reply:
Hmm. The URL is saying we're getting metadata for Dataset `drb`. So the "dataset" parameter is present, but the "metadata" parameter (which we expect to find in the JSON body) is missing.

There are probably several ways we could in theory be more specific. I'm working right now on generalizing and encapsulating the schema validation to understand the differences between `GET` and `PUT`, and between "URI template parameters" (dataset in this case), "URI query parameters" (`?metadata` on the `GET`) and "JSON body keys" (e.g., `metadata` here).

But I'm not sure I want to even try to parameterize the low-level schema validators enough to clarify the context of each error... it'd mean catching and re-writing the low-level validation errors at the higher level. It'd be more clear; is it worth the cost and complexity? My instinct is no since this is "developer" interfaces not "end user" interfaces... but the line can be thin.

In any case, I definitely don't want to attack that morass in *this* PR, so for now I'm good with this and I assume your approval means that you can live with it, too.